File: alds15c/main.cpp
#include <iostream>
#include <cmath>
#include <cstdio>
//#include <list>
#include <initializer_list>
#include <vector>
struct Point {
  double x,y;
};
/*
 * 1.与えられた線分を三等分にして、点3,点5を求める
 * 2.点3を原点に点5を回転させて点4を求める
 */

const double rad60 = M_PI/3;
const double cos60 = std::cos(rad60);
const double sin60 = std::sin(rad60);

Point* koch_curve(const Point p1, const Point p2, Point p[])
{
  Point tp1;
  tp1.x = (p2.x - p1.x)/3;
  tp1.y = (p2.y - p1.y)/3;
  p[0].x = p1.x + tp1.x;
  p[0].y = p1.y + tp1.y;

  p[2].x = p1.x + tp1.x *2;
  p[2].y = p1.y + tp1.y *2;

  p[1].x = p[0].x+ tp1.x * cos60 - tp1.y *sin60;
  p[1].y = p[0].y +tp1.x * sin60 + tp1.y *cos60;

  return p;
}

void koch_curve_r(const std::vector<Point>& ppl, const int depth, const int target)
{
  if (depth > target) return;

  const std::vector<Point>::const_iterator itEnd = ppl.end(); 
  std::vector<Point>::const_iterator it, itp1;

  for(it = itp1 = ppl.begin(), itp1++; itp1 != itEnd; it = itp1, itp1++)
  {
    Point p[3];
    koch_curve(*it, *itp1, p);
    std::initializer_list<Point> tlist = {p[0],p[1],p[2]};
    ppl.insert(itp1, tlist);
  }
  
  koch_curve_r(ppl, depth+1, target);
}

void print_koch_curve(std::vector<Point>& pl)
{
  const std::vector<Point>::const_iterator itEnd = pl.end();
  for(std::vector<Point>::const_iterator it = pl.begin(); it != itEnd; it++)
  {
    std::cout << (*it).x << " "<<(*it).y << std::endl;
  }
}

void init_list(std::vector<Point>& pl)
{
  Point p1=Point{0.0, 0.0}, p2 = Point{100.0, 0};
  pl.push_back(p1);
  pl.push_back(p2);
}

int main(void)
{
  std::vector<Point> pl;
  init_list(pl);
  

  print_koch_curve(pl);

  
}

File: alds15c/main.py
import numpy as np

def get_input():
    depth = int(input())
    return depth


def koch_curve(p1, p2):
    point1 = np.array(p1)
    point2 = np.array(p2)
    dp=(point2-point1)/3
    
    point3=point1+dp
    point5=point1+dp*2

    rad60 = np.pi/3
    
    rotarray = np.array(((np.cos(rad60), -np.sin(rad60)),(np.sin(rad60),np.cos(rad60))))
    dp2 = rotarray*dp
    dp2 = dp2[:,0]+dp2[:,1]
    point4 = point3 + dp2

    return ([point3.tolist(), point4.tolist(), point5.tolist()])



def rec_koch_curve(pl, depth, target):
    if (target <= depth) :
        return
    
    for i in range(len(pl)-1)[::-1]:
        tplist = koch_curve(pl[i] , pl[i+1])
        for p in tplist[::-1]:
            pl.insert(i+1, p)

    rec_koch_curve(pl, depth+1, target)

def print_curve(pl):
    for p in pl:
        print(p[0],p[1],sep=" ")

pl = [[0,0],[100,0]]
rec_koch_curve(pl, 0, get_input())
print_curve(pl)
